src/analysis/calculus/local_extr.lean

```diff
@@ -0,0 +1,293 @@
+/-
+Copyright (c) 2019 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+
+import topology.local_extr analysis.calculus.deriv
+
+/-! # Local extrema of smooth functions
+
+## Main definitions
+
+In a real normed space `E` we define `pos_tangent_cone_at (s : set E) (x : E)`.
+This would be the same as `tangent_cone_at ℝ≥0 s x` if we had a theory of normed semifields.
+This set is used in the proof of Fermat's Theorem (see below), and can be used to formalize
+[Lagrange multipliers](https://en.wikipedia.org/wiki/Lagrange_multiplier) and/or
+[Karush–Kuhn–Tucker conditions](https://en.wikipedia.org/wiki/Karush–Kuhn–Tucker_conditions).
+
+## Main statements
+
+First we prove that `0 ≤ f' y` whenever `a` is a local maximum of `f` on `s`,
+`f` has derivative `f'` at `a` within `s`, and `y` belongs to the positive tangent cone
+of `s` at `a`. Hence if both `y` and `-y` belong to the positive tangent cone, then `f' y = 0`.
+These facts are used to prove Fermat's Theorem: the derivative of a differentiable function
+at a local extremum point equals zero.
+
+Then we use Fermat's Theorem to prove Rolle's Theorem: given a function `f` continuous on `[a, b]`
+and differentiable on `(a, b)`, there exists `c ∈ (a, b)` such that `f' c = 0`.
+
+## Implementation notes
+
+For each mathematical fact we prove several versions of its formalization:
+
+* for maximums and minimums;
+* using `has_fderiv*`/`has_deriv*` or `fderiv*`/`deriv*`.
+
+For the `fderiv*`/`deriv*` versions we omit the differentiability condition whenever it is possible
+due to the fact that `fderiv` and `deriv` are defined to be zero for non-differentiable functions.
+-/
+
+universes u v
+
+open filter set
+open_locale topological_space classical
+
+section vector_space
+
+variables {E : Type u} [normed_group E] [normed_space ℝ E] {f : E → ℝ} {a : E}
+  {f' : E →L[ℝ] ℝ}
+
+/-- "Positive" tangent cone to `s` at `x`; the definition differs from `tangent_cone_at`
+by the requirement. One can think about `pos_tangent_cone_at` as `tangent_cone_at nnreal`
```

Comment:
You say `by the requirement`, but it looks like the sentence is not finished. Add `that `c` tends to `+∞`, instead of the fact that its norm tends to `+∞``.

Reply:
Fixed. Thank you for catching this!